src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs

```diff
@@ -129,15 +129,27 @@ public async Task StreamPool_MultipleStreamsInSequence_PooledStreamReused()
             // Stream has been returned to the pool
             Assert.Equal(1, _connection.StreamPool.Count);
 
-            await StartStreamAsync(3, _browserRequestHeaders, endStream: true);
+            await StartStreamAsync(3, _browserRequestHeaders, endStream: false);
+
+            // New stream has been taken from the pool
+            Assert.Equal(0, _connection.StreamPool.Count);
 
+            await SendDataAsync(3, _helloBytes, endStream: true);
             await ExpectAsync(Http2FrameType.HEADERS,
-                withLength: 55,
-                withFlags: (byte)(Http2HeadersFrameFlags.END_HEADERS | Http2HeadersFrameFlags.END_STREAM),
+                withLength: 37,
+                withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
+                withStreamId: 3);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 5,
+                withFlags: (byte)Http2DataFrameFlags.NONE,
+                withStreamId: 3);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 0,
+                withFlags: (byte)Http2DataFrameFlags.END_STREAM,
                 withStreamId: 3);
 
-            // New stream has been taken from the pool
-            Assert.Equal(0, _connection.StreamPool.Count);
+            // Yield to let Kestrel finish moving the stream to the completed queue
+            await Task.Yield();
```

Comment:
I know inline scheduling can make these tests a little weird, but `await Task.Yield();` rarely completely fixes flakiness.

Even if we're just trying to let Http2Connection's frame processing loop continue so UpdateCompletedStreams can run, there's no guarantee that the test method doesn't continue on another thread before UpdateCompletedStreams actually runs. It's entirely possible that the frame processing loop gets preempted and doesn't continue any further by the time the test method continues.

And even though this test calls SendPingAsync and awaits the ping back to "trigger the stream to be returned to the pool", there's no guarantee that the ping doesn't get received and responded to before UpdateCompletedStreams is even called.

We might need to make the stream pool virtual so we can replace it in the tests and properly deflake them with TCSs. Another hack might be to exit the `while (Http2FrameReader.TryReadFrame(` loop after a ping is received. We'd have to be extra careful to continue the outer loop even if we get a completed ReadResult in that case if there's more data left in the connection's PipeReader. Or maybe we can verify the pool still contains one stream after the connection is fully stopped.

Reply:
Talked offline. Issue is different to what Stephen thought it was. I have replaced the yield with TCS in the app delegate.